File: CompressibleAerodynamicCalculator.cpp
#include<iostream>
#include<cmath>
#include <limits>
using namespace std;

class IsentropicFlowRelations{
public:
    float gamma;
    float Mach;

    IsentropicFlowRelations(float gamma, float Mach){
        this->gamma = gamma;
        this->Mach = Mach;
    }

    float staticByTotalPressure(){
        return pow(1 +  (gamma - 1)/2 * pow(Mach,2), -((gamma)/(gamma-1)));
    }

    float staticByTotalTemperature(){
        return pow(1 +  (gamma - 1)/2 * pow(Mach, 2), -1);
    }

    float staticByTotalDensity(){
        return pow(1 +  (gamma - 1)/2 * pow(Mach,2), (-1/(gamma-1)));
    }

    float staticByChokeArea(){
        return  pow(1 + pow(Mach,2) * (gamma - 1)/2 , (gamma + 1)/(2*(gamma-1))) * pow((gamma + 1)/2, -((gamma + 1)/(2*(gamma -1))))/Mach;
    }

    float machAngle(){
        cout<<"Mach Angle is in Degree."<<endl;
        return asin(1/Mach)* (180/M_PI);
    }

    float prandtlMeyerAngle(){
        float a = sqrt((gamma + 1) / (gamma - 1)) * atan(sqrt((gamma - 1) / (gamma + 1) * (pow(Mach,2)- 1)));
        float b = atan(sqrt(pow(Mach,2) - 1));
        return (a - b) * (180 / M_PI);
    }

};

class NormalShockRelations : public IsentropicFlowRelations{
public:

    NormalShockRelations(float gamma, float Mach) : IsentropicFlowRelations(gamma, Mach){}
    float secondMach() {
        float a = ((gamma - 1) * pow(Mach,2)) + 2;
        float b = 2 * gamma * pow(Mach,2) - (gamma - 1);
        return sqrt(a / b);
    }

    float AfterDensity() {
        return ((gamma + 1) * pow(Mach,2)) / ((gamma - 1) *pow(Mach,2) + 2);
    }

    float AfterTemperature() {
        float a = (2 * pow(Mach,2) * gamma - (gamma - 1)) * ((gamma - 1) * pow(Mach,2) + 2);
        float b = (gamma + 1) * (gamma + 1) * pow(Mach,2);
        return a / b;
    }

    float AfterPressure() {
        float a = 2 * gamma * pow(Mach,2) - (gamma - 1);
        float b = gamma + 1;
        return a / b;
    }

    float AfterTotalPressure() {
        float a = pow(AfterDensity(), gamma / (gamma - 1));
        float b = pow(1 / AfterPressure(), 1 / (gamma - 1));
        return a * b;
    }
};

class ObliqueShockRelations : public IsentropicFlowRelations{
public:
    float turnAngle;

    ObliqueShockRelations(float gamma, float Mach, float turnAngle) : IsentropicFlowRelations(gamma, Mach) {
        this->turnAngle = turnAngle;
    }

    float theta_beta_m_relation(float beta) {
        // Theta-Beta-Mach relation equation.
        float beta_rad = beta * M_PI / 180.0;
        float term1 = 2 * (1 / tan(beta_rad)) * (Mach * Mach * pow(sin(beta_rad), 2) - 1);
        float term2 = Mach * Mach * (gamma + cos(2 * beta_rad)) + 2;
        return tan(turnAngle * M_PI / 180.0) - (term1 / term2);
    }

    float theta_beta_m_relation_derivative(float beta) {
        // Derivative of Theta-Beta-Mach relation equation with respect to beta.
        float beta_rad = beta * M_PI / 180.0;
        float term1 = 2 * (1 / tan(beta_rad)) * (Mach * Mach * pow(sin(beta_rad), 2) - 1);
        float term2 = Mach * Mach * (gamma + cos(2 * beta_rad)) + 2;
        float derivative_term1 = -2 * (1 / (sin(beta_rad) * sin(beta_rad))) * (Mach * Mach * pow(sin(beta_rad), 2) - 1);
        float derivative_term2 = 4 * Mach * Mach * sin(beta_rad) * cos(beta_rad);
        return -(derivative_term1 + term2 * derivative_term2) / (term2 * term2);
    }

    float find_shock_angle() {
        // Initial guess in degree
        float beta_guess = 45.0;
        float beta_solution = beta_guess;
        float tolerance = 1e-2;
        int max_iterations = 10;
        int iteration = 0;

        while (fabs(theta_beta_m_relation(beta_solution)) > tolerance && iteration < max_iterations) {
            beta_solution -= theta_beta_m_relation(beta_solution) / theta_beta_m_relation_derivative(beta_solution);
            iteration++;
        }

        if (iteration == max_iterations) {
            cout << "Solution did not converge" << endl;
            return std::numeric_limits<float>::quiet_NaN();
        }

        return beta_solution;
    }

    float deflection_angle() {
        float beta = find_shock_angle();
        return turnAngle;
    }

    float changeInDensity() {
        float beta_rad = find_shock_angle() * M_PI / 180.0;
        float a = (gamma + 1) * Mach * Mach * pow(sin(beta_rad), 2);
        float b = (gamma - 1) * Mach * Mach * pow(sin(beta_rad), 2) + 2;
        return a / b;
    }

    float changeInPressure() {
        float beta_rad = find_shock_angle() * M_PI / 180.0;
        float a = 2 * gamma * Mach * Mach * pow(sin(beta_rad), 2) - (gamma - 1);
        return a / (gamma + 1);
    }

    float changeInTemperature() {
        return changeInPressure() * (1 / changeInDensity());
    }

    float totalPressureChange() {
        float a = pow(changeInDensity(), gamma / (gamma - 1));
        float b = pow(1 / changeInPressure(), 1 / (gamma - 1));
        return a * b;
    }

    float AfterShockMachNumber() {
        float beta_rad = find_shock_angle() * M_PI / 180.0;
        float a = 2 + (gamma - 1) * Mach * Mach * pow(sin(beta_rad), 2);
        float b = 2 * gamma * Mach * Mach * pow(sin(beta_rad), 2) - (gamma - 1);
        float c = sqrt(a / b);
        return (1 / (sin(beta_rad - (turnAngle * M_PI / 180.0)))) * (c);
    }
};

int main() {
    IsentropicFlowRelations p1(1.4, 2);
    cout << "P/Po: " << p1.staticByTotalPressure() << endl;
    cout << "T/To: " << p1.staticByTotalTemperature() << endl;
    cout << "rho/rho_o: " << p1.staticByTotalDensity() << endl;
    cout << "A/A*: " << p1.staticByChokeArea() << endl;
    cout << "Mach Angle: " << p1.machAngle() << endl;
    cout << "P-M Angle: " << p1.prandtlMeyerAngle() << endl;

    NormalShockRelations p2(1.4, 2);
    cout << "After Normal Shock M: " << p2.secondMach() << endl;
    cout << "Density change: " << p2.AfterDensity() << endl;
    cout << "Temperature change: " << p2.AfterTemperature() << endl;
    cout << "Pressure change: " << p2.AfterPressure() << endl;
    cout << "Total Pressure change: " << p2.AfterTotalPressure() << endl;

    ObliqueShockRelations p3(1.4, 2, 20);
    cout << "Shock Angle: " << p3.find_shock_angle() << endl;
    cout << "Change in Density: " << p3.changeInDensity() << endl;
    cout << "Change in Pressure: " << p3.changeInPressure() << endl;
    cout << "Change in Temperature: " << p3.changeInTemperature() << endl;
    cout << "Total Pressure Change: " << p3.totalPressureChange() << endl;
    cout << "After Shock Mach Number: " << p3.AfterShockMachNumber() << endl;

    return 0;
}

File: CompressibleAerodynamicCalculator.py
import numpy as np
import math
from math import sqrt, sin, cos, tan, atan
from scipy.optimize import fsolve

class IsentropicFlowRelations():
    def __init__(self, gamma, R, Mach):
        self.gamma = gamma
        self.R = R
        self.Mach = Mach

    def staticByTotalPressure(self):
        return pow(1 +  (self.gamma - 1)/2 * self.Mach**2, -(self.gamma/(self.gamma-1)))

    def staticByTotalTemperature(self):
        return pow(1 +  (self.gamma - 1)/2 * self.Mach**2, -1)

    def staticByTotalDensity(self):
        return pow(1 +  (self.gamma - 1)/2 * self.Mach**2, (-1/(self.gamma-1)))

    def staticByChokeArea(self):
        return  pow(1 + self.Mach**2 * (self.gamma - 1)/2 , (self.gamma + 1)/(2*(self.gamma-1))) * pow((self.gamma + 1)/2, -((self.gamma + 1)/(2*(self.gamma -1))))/self.Mach

    def machAngle(self): #degree
        print("Mach Angle is in Degree.")
        return math.asin(1/self.Mach)* (180/math.pi)

    def prandtlMeyerAngle(self):
        a = sqrt((self.gamma + 1) / (self.gamma - 1)) * math.atan(sqrt((self.gamma - 1) / (self.gamma + 1) * (self.Mach**2 - 1)))
        b = math.atan(sqrt(self.Mach**2 - 1))
        return (a - b) * (180 / math.pi)

class NormalShockRelations(IsentropicFlowRelations):
    def secondMach(self):
        a = ((self.gamma-1)* self.Mach**2) + 2
        b = 2*self.gamma* self.Mach**2 - (self.gamma-1)
        return sqrt(a/b)

    def AfterDensity(self):
        return ((self.gamma + 1)* self.Mach**2)/ ((self.gamma-1)* self.Mach**2 + 2)

    def AfterTemperature(self):
        a = (2* self.Mach**2* self.gamma - (self.gamma -1)) * ((self.gamma-1)* self.Mach**2 + 2)
        b = (self.gamma + 1)**2 * self.Mach**2
        return a/b
    def AfterPressure(self):
        a = 2*self.gamma*self.Mach**2 - (self.gamma - 1)
        b = self.gamma + 1
        return a/b

    def AfterTotalPressure(self):
        a = pow(self.AfterDensity(), self.gamma/(self.gamma -1))
        b = pow(1/self.AfterPressure(), 1/(self.gamma -1))
        return a*b

class ObliqueShockRelations(IsentropicFlowRelations):
    def __init__(self,gamma, R, Mach, turnAngle):
        super().__init__(gamma, R, Mach)
        # converting the degree into Radians
        self.turnAngle = np.deg2rad(turnAngle)

    def theta_beta_m_relation(self, beta):
        """Theta-Beta-Mach relation equation."""
        beta = np.deg2rad(beta)
        term1 = 2 * (1 / np.tan(beta)) * (self.Mach**2 * np.sin(beta)**2 - 1)
        term2 = self.Mach**2 * (self.gamma + np.cos(2 * beta)) + 2
        return np.tan(self.turnAngle) - (term1 / term2)

    #TODO Find shock angle (beta) using numerical solver.
    def find_shock_angle(self):
        # Initial guess in degree
        beta_guess = 45
        beta_solution = fsolve(self.theta_beta_m_relation, beta_guess)
        return beta_solution[0]

    def deflection_angle(self):
        beta = self.find_shock_angle()
        return np.rad2deg(self.turnAngle)

    def changeInDensity(self):
        a = (self.gamma + 1)* self.Mach**2* sin(np.deg2rad(self.find_shock_angle()))**2
        b = (self.gamma - 1)* self.Mach**2* sin(np.deg2rad(self.find_shock_angle()))**2 + 2
        return (a/b)

    def changeInPressure(self):
        a = 2* self.gamma * self.Mach**2 * sin(np.deg2rad(self.find_shock_angle()))**2 - (self.gamma - 1)
        return a/(self.gamma + 1)

    def changeInTemperature(self):
        return self.changeInPressure() * (1/self.changeInDensity())

    def totalPressureChange(self):
        a = pow(self.changeInDensity(), self.gamma/(self.gamma - 1))
        b = pow(1/self.changeInPressure(), 1/(self.gamma - 1))
        return a*b

    def AfterShockMachNumber(self):
        a = 2 + (self.gamma -1) * self.Mach**2 * (sin(np.deg2rad(self.find_shock_angle())))**2
        b = 2 * self.gamma * self.Mach**2 * (sin(np.deg2rad(self.find_shock_angle())))**2 - (self.gamma -1)
        c = np.sqrt(a/b)
        return (1/(sin(np.deg2rad(self.find_shock_angle()) - (self.turnAngle)))) * (c)


p1 = IsentropicFlowRelations(1.4, 343.24, 2)
p2 = NormalShockRelations(1.4, 343.24, 2)
p3 = ObliqueShockRelations(1.4, 343.24, 2, 20)
print("P/Po: %f" % p1.staticByTotalPressure())
print("T/To: {}".format(p1.staticByTotalTemperature()))
print("rho/rho_o: %f" % p1.staticByTotalDensity())
print("A/A*: %f" % p1.staticByChokeArea())
print("Mach Angle: %f" % p1.machAngle())
print("P-M Angle: %f" % p1.prandtlMeyerAngle())
print("After Normal Shock M: %f" % p2.secondMach())
print("Density change: %f" % p2.AfterDensity())
print("Temperature change: %f" % p2.AfterTemperature())
print("Pressure change: %f" % p2.AfterPressure())
print("Total Pressure change: %f" % p2.AfterTotalPressure())
print("Shock Angle: %f" % p3.find_shock_angle())
print("Change in Density: %f" % p3.changeInDensity())
print("Change in Pressure: %f" %p3.changeInPressure())
print("Change in Temperature: %f" % p3.changeInTemperature())
print("Total Pressure Change: %f" % p3.totalPressureChange())
print("After Shock Mach Number: %f" % p3.AfterShockMachNumber())


